README.md
# Santa Wrap

This article is part of [Advent of Haskell](https://adventofhaskell.com/) 2020.
Make sure to check the other articles too!

## Our Christmas project

Year 2020 is terrible. Santa's Workshop had to take stay-at-home restrictions.
However, Elves want to do the utmost to keep Christmas' spirit alive. This
year, Elves will work from home rather than work in the streamlined workshop.
They will get back to a more traditional way of preparing presents, like
Elderly Elves who live more than 500 years, remember.  In short, every Elf
works in their own garage. They receive the Christmas wishlists of neighboring
families: crafting toys, sewing clothes, or cooking delicious cookies.

One particularly annoying task for work-from-home Elves, however, is to wrap
presents. They had avoided this task in the recent years thanks to automation.
Elves all have different pieces of wrapping-paper at home, in particular
wrapping-paper rolls vary in ratio and vary in size from one Elf to another
one. Elves want some software to help them cut wrapping-paper for the
presents they prepare. Some Elves have tools to cut paper (think of a
cookie-cutter, but for large sheets of paper), whereas others only have
cissors. Wrapping may become scarce if used inefficiently, and anyway ordering
takes time. Thus, they would like to have a rough idea of how to efficiently
cut the paper rolls given their own specific constraints. It does not have to
be perfect, but it cannot be completely naive either. Thus, Elves need some
creative idea to save Christmas' spirit. For the better or for the worst
Elves resorted to use software.

Fortunately, an Elf has learnt Haskell, a language especially good for building
small command line tools and with libraries to generate images. He also
has learnt MiniZinc, a language especially good for solving this family of
problems. Put together, these languages can help building a small tool that
Elves can run at home with their own input. This computer-savvy Elf suggests
the following tradeoff:
- Elves will input the quantity and shape of the wrapping for the presents (or
  the shape of their cookie-cutter tools that fit for the present)
- the shape of the wrapping is blocky, that is, made of various rectangle
  blocks, rather than arbitrarily shaped (it simplifies cutting and folding
  paper with rulers and guides anyway)
- Elves will also input the size of their own paper sheets
- the output is a footprint image of how to cut the paper to fit all present
  (when possible)
Some Elves expected more specialized tools, some cared especially about saving
wrapping paper. However, time is running out, our software Elf-ineer only has a
day to build this software or it will be too late to save Christmas.

I am not sure I should reveal this, but, I work in Santa Incorporated oh oh oh.
Do not tell your kids! in exchange I'll explain you how I built `santa-wrap` in
a rush to help the Elf. Thus, the main goal for this article is to relate how
to iterate quickly on a practical one short day project. Secondary goals are to
pike your interest about MiniZinc and introduce you a library named
`minizinc-process`.

## Approach

The type of problem for cutting wrapping paper is known as a Packing problem.
These problems are generally difficult to solve efficiently. However, for small
sizes or particular form of constraints, some solutions could be efficiently
found (or the litterature may have some heuristics that work well empirically).
Generally, the difficulty comes from integrality of constraints, that is you
either can wrap a present or not wrap the present, you cannot wrap a fraction
of a present. Such problems come up a lot in industrial settings where some
resource is scarce. In this particular case, we cheated a bit as we came with
an idea to write a blog post. However, with a bit of practice you will
recognize packing problems everywhere.

We want to show off how to combine MiniZinc (a mathematical optimization
language and toolchain) with Haskell (which readers of this Advent of Haskell
submission are familiar with). We will show the progression of a tool, with
building a value-adding prototype in a day or two of work for some Elf familiar
with Haskell and Minizinc.

## Motivating our Iterations

The author built this tool in four broad iterations: implement a demo solver
in MiniZinc, generate simple pictures for each wrapping, introduce the concept
of layout, and, bridge the solver. We motivate each iteration before jumping
into lengthier explanations.

*Implemeting a solver in MiniZinc*: I wanted to leverage powerful global
constraints with their own heuristics/optimized solutions that exist in
MiniZinc. Such a model helps us formalize the problem in better terms and
understand what are inputs and what are outputs. In a business settings, such
prototypes allow to raise the important questions. In that case, the whole
problem is solved with a single [global contraint](https://www.minizinc.org/doc-2.3.0/en/lib-globals.html#packing-constraints) named `geost_bb`. Knowing
which constraints help modeling a problem is a type of expertise that takes
time to acquire but that can change drastically the viability of such a tool.

*Generate simple pictures*: our tool needs to generate pictures of the
wrapping, and I find a blog-post with picture more appealing than without.
Visualizations illustrate the problem. I am not familiar with low-level
graphics libraries. I typically use high-level tools for my charts like [ggplot
in R](https://ggplot2.tidyverse.org/) -- they typically get in the way if what
you want is pixel-perfect position of image width/height. For this exercise,
I found [JuicyPixels](https://hackage.haskell.org/package/JuicyPixels) and I
need to validate quickly that JuicyPixels will fit the needs.  This library is
appealing because I could generate PNGs with little dependencies. I had some
experience with [Gloss](https://hackage.haskell.org/package/gloss). However, I
discarded this choice for I only need static images and not animations (and
Gloss has way more dependencies than JuicyPixels).

*Introduce the concept of layout*: pictures become more interesting when we
start to have multiple boxes. Once I know how to print a layout I am done with
the graphical aspects of the problem. Minor improvements can still come later
but it is mind-relaxing to call something done. Further, at this point it is
easy to show some naive layouts.

*Bridge the solver*: this is the boiler-plate part of the work.
`minizinc-process` actually have a TemplateHaskell helper to generate an input
and output representations but this approach is limited: it cannot work with
sets in its current version and does not preclude writing boilerplate to
transform between "business-domain" representations and solver input/output
representations. Plus, the goal of the article is to illustrate the pro/cons.
This part of boiler-plate is the one requiring the most care, hence my usage of
newtypes to differentiate indices.

Graphically, the following dependency graph shows the various steps I had in
mind.  ![img](images/deps.png) In red are the two chunks of work which I
assessed has being the most uncertain (i.e., requiring the most discovery) or
most meticulous (i.e., spending time building accurate types to reduce the risk
of bugs). 

Let's discuss how this four iterations looked like.

## The four iterations

We now dive in more details into each iteration. Recall that I time-boxed this
project to the equivalent of a short-day-of-work. Readers should be able to
read these independently.

### A solver in MiniZinc

We basically call the _global contstraint_ `geost_bb`. Let's unpack how it
works, starting from a small MiniZinc overview and a mini tutorial on
`geost_bb`.

In [MiniZinc](https://www.minizinc.org/) you separate the structure of a
problem (which we call a model) and its particular instance (determined by
input parameters). Thus a model can run for various input parameters. Further,
the model separate two families of variables: the _par_ variables (parameters
that are given or can be deterministically computed from other parameters) and
_var_ variables (decisions that the solver will make). Constraints are
statement that link variables together. The solver, then works out solutions
(that is assignments of variables) that are consistent with all constraints. It
is typically easy to verify that a given assignment matches all constraints.
For instance, in our Santa-Wrap tool can visually inspect whether or not
cutting guides for different present intersect. Whereas, it is typically hard
to find a given assignment for all possible _par_ inputs (or an optimal one
when a criteria of goodness exists).

In Haskell parlance: `solver :: Model -> Parameters -> Maybe Decisions`, which,
partially-applied for some `Model` would give `solveProblem :: Parameters ->
Maybe Decisions`. To encode this faithfully in Haskell we would need to ensure
that the `Parameters` and `Decisions` types actually depends on the `Model`;
somehow, discussing these type-level programming techniques is out of scope for
this article.

Typically, writing MiniZinc is a bit like SQL queries: you describe what you
want and let the engine fetch it for you. The ability to summon a magic tool to
solve our wishes comes with some limitations: expressivity is limited, and you
need to be careful what you wish for. For instance, when writing MiniZinc you
need to discuss about finite inputs/outputs (e.g, no unbounded recursion).
Further, small changes on the structure of the problem can drastically change
performance (like switching from a index scan to a table scan in SQL query
plan).

The comparison with SQL does not stops at the declarative aspect. Typical
domain models in Haskell have collections such as
[Lists](https://hackage.haskell.org/package/base-4.14.1.0/docs/Data-List.html)
or
[Maps](https://hackage.haskell.org/package/containers-0.6.4.1/docs/Data-Map.html)
or custom [recursively-shaped objects](http://www.cse.chalmers.se/edu/year/2018/course/TDA452/lectures/RecursiveDataTypes.html) etc. Once mapped into tabular form in SQL, special
conversion care is required.  For instance, in Haskell, a type `User { age ::
Int , height :: Int }` would map to a set of integers (one user-id per user)
and two arrays indexed by user-id. So that Haskell domain objects like these.

```haskell
users :: [ User ]
users = [ User 23 172, User 45 180, User 36 165 ]
```

Would translate to the following MiniZinc.

```mzn
set of int: USER = 1..3;
array[USER] of int: age = [ 23, 45, 36 ];
array[USER] of int: height = [ 172, 180, 165 ];
```

Readers familiar with Relational Algebra, DataParallel Haskell, Columnar
Storage, or Entity Component System will find comonalities. We will work this
out for our problem in more detail in the last iteration. So now let's look at
`geost_bb`.

Here is a snapshot of `geost_bb`
[documentation](https://www.minizinc.org/doc-2.5.0/en/lib-globals.html)
![img](images/geost-doc.png). Which sorts of packs blocky-structure on a
multi-dimensional space. This constraint does not try to shrink gaps between
objects, it "merely" (I don't want to downplay how great this is) force
assignments of variables such that no overlap exists between the packed
objects.

We annotate the parameters below to explain why using this function make sense
in Santa-Wrap.  Besides the opaque name, it does exactly what we want to help
the Elves and even a bit more.

- *k*: here we live in two dimensions, that an easy parameter, we'll hardcode it
- *rect_size*: size of each rectangles along each dimension
- *rect_offset*: offset of rectangles long each dimension
- *shape*: wrapping shapes are built from a set of a rectangles, for instance presents
  we can wrap in a simple rectangle would have a single rectangle. More
  elaborated wrapping shapes may require two or more adjacent rectangles.
  For instance the sword here is built with two overlapping rectangles:
  ![img](images/sword.png) a rectangle is horizontal (largest _rect_size_ is the 1st dimension) and another rectangle is vertical (largest _rect_size_ is the 2nd dimension) and both rectangles are translated along an axis (i.e., one non-zero _rect_offset_). Multiple shapes could re-use the same rectangles if
  they happen to match (hence, saving a bit on the complexity -- but we won't go into this   level of sophistication).
- *l* and *u*: upper and lower bounds for each wrapping, in our case all
  objects can equally be positioned on the wrapping, hence the lower bound is 0
  and the upper bound the size of the roll (minus one if you care about
  millimeters and took no margins of error at all in your measurements)

Then `geost_bb` then contributes constraints to tell the solver to make a
number of decisions:

- `x`: the coordinates of the _wrapping shapes_ (which are themselves built of
  rectangular blocks). For our sword example, we need to add the `rect_offset`
  to the `x` coordinate of the shape to position a rectangle.
- `kind`: the shape used by each object. This mechanism allows `geost_bb`
  to chose any shape for the different objects. External constraints actually prevent
  `geost_bb` does not know how many times it must be using the wrapping-shape
  for a sword, and unconstrained the solver would rather prefer the simpler
  wrapping-shape for a book. This mechanism actually is a feature: the freedom
  to pick among shapes allows to model rotation of shapes, we will not use this
  for now.

In our case, for a first version of the tool we do not need to make that many
decisions: only `x` above is required. Declaring all inputs to give to
`geost_bb` is required, however. We still need to actually declare our `par`
and `var` variables. 

We will have to discuss about a number of entities:
- two dimensions: `int: nDims = 2; set of int: DIM = 1..nDims`
- wrapping shapes:  `int: nShapes; set of int: SHAPE = 1..nShapes`
- rectangular blocks: `int: nBlocks; set of int: BLOCK = 1..nBlocks`
- each object we need to wrap: `int: nObjs; set of int: OBJ = 1..nObjs`
- the roll: it is implicit that we cut one roll, so we do not need to declare anything

Then, some characteristics of entities:
- the size of the roll: `int: x_max; int: y_max`
- the size of individual rectangular blocks and their offsets in each dimension: `array[BLOCK,DIM] of int: rect_sizes; array[BLOCK,DIM] of int: rect_offs`
- the coordinates of where to position cuts for our solution: `array[OBJ,DIM] of var int: coordinates`

Finally there is relational information linking various entities.
- to each object we associate a given shape (present1 is a Guitar, present2 is a book etc.): `array[OBJ] of SHAPE: object_shapes;`
- each shape is the collage of a set of rectangle, we could use various
  external representation to model the set of rectangles (e.g., an
  `array[SHAPE,RECTANGLE] of bool` is isomorphic to an `array[SHAPE] of set of
  RECTANGLE`) but we will stick to the `set` representation that `geost_bb`
  requires (using  another input would require extra constraints to link input
  and `geost_bb` representations)

All in all, the model is in `models/santa-wrap.mzn` and an example input for it is `models/santa-wrap001.dzn`. You can run it with `minizinc models/santa-wrap.mzn models/santa-wrap001.dzn`. And the output I got was:

```
coordinates = array2d(1..10, 1..2, [8, 0, 0, 6, 7, 0, 0, 5, 6, 0, 0, 4, 4, 0, 0, 2, 2, 0, 0, 0]);
```

Which you can read as "first object" is a (8,0), the "second object" is at
(0,6) etc. Elves could work with such an output, however a visual display is
better. Plus they want to prepare their cutting boards, build guides, or simply
pass an eye-check that things are in the right place.

### Generate simple pictures

I have had little interest in graphics in the past. Thus I vaguely know some
Haskell libraries exist like
[diagrams](https://hackage.haskell.org/package/diagrams), `gloss`,
`JuicyPixels`, or [reanimate](https://hackage.haskell.org/package/reanimate).
For this project, I had to dive a bit more and actually pick some library. My
choice settled on `JuicyPixel` for two main reasons: on the first hand, this
library has only a small amount of dependencies (especially system dependencies
-- `gloss` requires some OpenGL). On the other hand, the documentation was
welcoming: I looked for `IO ()` because I looked for some "main" or "print"
equivalent. The first hit on the top module is `generateImage`.  No kidding, here
is the screenshot to prove it: first hit ![img](images/haddock-jp.png).

The type information tells me the library actually scans every pixel one by
one. It is very likely super inefficient. That said, the cost of optimizing the
packing is the major bottleneck in this problem. Elves can wait a minute to
save hours. A warming case that does not corroborate the claim that Haskell
libraries are poorly documente. My first action was to verify that the example
actually works.

```shell
cabal repl
> import Code.Picture
> let path = "images/toto.png"
> let pixelRenderer x y = PixelRGB8 (fromIntegral x) (fromIntegral y)
> writePng path $ generateImage pixelRenderer 250 300
```

And voila! ![img](images/toto.png) thank you so much `JuicyPixels`. It's not
"too good to be true": the example works.

Having secured that I can generate PNG, what is left? we need to define some
Haskell data types to map loosely with what I've learnt is required in the
MiniZinc optimization. However we'll be using Haskell records rather than
columnar values. The idea here is to use these data types as main Haskell types
for parsing a format that Elves are likely to know such as JSON. Further we
want to allow Elves to pass extra information like quantities of each present
they will wrap, colors and names for pictures before crafting the guides etc.

Some excerpt:

```haskell
data Wrapping = Wrapping {
    name       :: Name
  , colorNum   :: Int
  , quantity   :: Quantity
  , rectangles :: [ Rectangle ]
  } deriving (Show, Generic)
instance FromJSON Wrapping

data Rectangle = Rectangle {
    rectX :: Length
  , rectY :: Length
  , offX  :: X
  , offY  :: Y
  } deriving (Show, Generic)
instance FromJSON Rectangle
```

Wrappings are built of Rectangles. Each Rectangle as length/width (lengths in
both dimensions `rectX`, `rectY`) as well as some offset in both dimension. In
addition, each Wrapping has a name, a quantity and some color to visualize more
easily a layout. But before speaking about Layouts, what we can do is to
generate pictures for each Wrapping. This way, Elves can calibrate the units
and verify individual length.

I need to adapt the `JuicyPixels` example I copy pasted to build a function
`printWrapping :: Wrapping -> IO()`. In short, it has a `Wrapping` and its
associated `Rectangles` in scope and we need to implement `pixelRenderer`.

So I'll build utility functions `findRectangle x y` which finds a rectangle
with the right sizes and offset to contain the pixel `(x,y)`.

A single pixel and rectangle test is

```
rectangleContainsPixel x y rect =
  let x0 = scale $ offX rect
      x1 = scale $ offX rect + rectX rect
      y0 = scale $ offY rect
      y1 = scale $ offY rect + rectY rect
  in  x0 <= x && x < x1 && y0 <= y && y < y1
```

Which checks for intervals in both X and Y. The `scale` function just
multiplies tenfold every magnitude (i.e., a dimension of 1 Length is shown with
10 pixel). With that utility function we can now test every rectangles: if some
is found, we color the Pixel with some value, otherwise we use default
background value.

```haskell
pixelRenderer x y =
  case findRectangle x y of
    Nothing -> PixelRGB8 (fromIntegral x `mod` 32) (fromIntegral y `mod` 32) 128
    Just idx -> pixelColor idx

findRectangle x y = List.findIndex (rectangleContainsPixel x y) (rectangles wrapping)
```

I use `List.findIndex` so that on a hit i can also attribute some color (see
`pixelColor :: Int -> PixelRGB8` which defines a palette) and paint the
rectangle in a different color.

Another utility function is needed to give `JuicyPixel` the size of the picture
to create.  This size corresponds to the farthest point in the set of off-set
rectangles `maxX = maximum (0 : [ rectX r + offX r | r <- rectangles wrapping
])`, I named this value the `extentX` of a `Wrapping` in later commits.  I use
comprehension list notation. I add a 0 so that `maximum` does not crash on an
empty list.

Our `main` function basically parses `stdin` into some `[Wrapping]` and
`traverse` this list with our PNG generator.

```haskell
mainParts = do
  wrappings <- eitherDecode' @[Wrapping] <$> ByteString.getContents
  case wrappings of
    Left err -> print err
    Right xs -> traverse_ printWrapping xs
  putStrLn "done!"
```

The parsing leverages the automatically-generated `instance FromJSON` from
`aeson` (extension `DeriveGeneric`).  Note that we use `TypeApplications` for
the syntax `@[Wrapping]` which explicitates what we are decoding from the
content in `stdin`, which also helps GHC provide less general error message
when I change and break things.

This way, our main binary can generate individual images.

```console
$ santa-wrap parts < input/wrapping001.json 
$ done!
```

The parts are individual `.png` files as this listing shows:
![img](images/parts.png).  Elves can use these parts to prepare some guides to
later cut wrapping paper rapidly.

Now let's finally introduce the concept of layout.

### Introduce the notion of layout

The layout is a collection of positioned wrappings. There's an obvious Functor
instance for Positioned. I could refactor Rectangle into a Positioned Rect.
This way, a Layout is a `[ Positioned [ Positioned Rect ] ]` and it is probably
collapsable into a single list by translating the inner Positioned objects.
Something like `flatten :: [ Positioned [ Positioned a ] ] -> [ Positioned a ]`
I feel tempted to actually dig in this direction. However, I'm far from done
and the clock ticks. In short, I will save this for later and perform a good
amount of copy-pasting.

The Haskell code is simply:

```haskell
data Positioned a = Positioned {
    posX :: X 
  , posY :: Y 
  , item :: a 
  } deriving (Show, Functor)
    
data Layout = Layout {
    positionedWrappings :: [ Positioned Wrapping ]
  } deriving (Show) 
```

And now my `printLayout` function has a very similar `pixelRenderer`.
Two distinctions though:
- to color a pixel you need to determines if a `Wrapping` has a `Rectangle`
  and the `Rectangles` are translated by `posX` and `posY`.
- we want to add a few pixels of `margin` so that a series of adjacent squares
  do not appear like a long rectangle, this small margin will ease verifying
  layouts at a glance.

Hence, our `rectangleContainsPixel` test takes all these things into account:

```haskell
rectangleContainsPixel :: X -> Y -> Int -> Int -> Rectangle -> Bool
rectangleContainsPixel posx posy x y rect =
        let x0 = scale $ posx + offX rect
            x1 = scale $ posx + offX rect + rectX rect
            y0 = scale $ posy + offY rect
            y1 = scale $ posy + offY rect + rectY rect
        in  (x0 + margin) <= x && x < (x1 - margin) && (y0 + margin) <= y && y < (y1 - margin)
```

How do you verify that the margin is pretty enough? Simply generate a naive
Layout function. We just lay out shapes left-to-right. 

```haskell
linearLayout :: [ Wrapping ] -> Layout
linearLayout wrappings = Layout $ List.zipWith (\w (x,y) -> Positioned x y w) wrappings xys
  where
    xys :: [(X,Y)]
    xys = List.scanl placeToTheLeft (0,0) wrappings 

    placeToTheLeft :: (X, Y) -> Wrapping -> (X, Y)
    placeToTheLeft (x,y) w = (x + extentX w, 0)
```

The interesting aspect is the use of `List.scanl` to perform a fold an
accumulate intermediary values.  The accumulator is the position for the next
`Wrapping` (hence, we start at `(0,0)` and then we move right by an amount
large enough to ensure there will be no overlap: what we named the _extent_ of
the wrapping shape.

This function does not even take care of the bounds of the roll.  This function
is merely useful to have some comparison point and to exercise the
`printLayout` function.

Running the following command will give the image below.

```console
$ santa-wrap linear-layout ./layout.png < input/wrapping003.json 
done!
```
![img](images/linear-layout.png)

Now we see the margins, the coloring. We also have a MiniZinc model for
non-naive layouts. Let's bring the two things together.

### Actually bridge the Haskell and MiniZinc code

Showcasing my `minizinc-process` library is, a main motivator (second to saving
Christmas' spirit, though).

Remember that, we want to approach ergonomics for the solver so that in
Haskell-land it feels like we have a function `solver :: Model -> Parameters ->
Maybe Decisions`.  Here enters `minizinc-process`, an helper library that
actually uses MiniZinc support for JSON input/outputs.  This is feasible with
`runLastMinizincJSON`, which has type:

```haskell
runLastMinizincJSON ::
  (ToJSON input, FromJSON answer) =>
  MiniZinc input answer ->
  input ->
  IO (Maybe answer)
runLastMinizincJSON = ...
```

The typeclass constraints on `input` and `output` allow to serialize back and
forth using JSON.  The `Minizinc input answer` actually contains the model
(e.g., a filepath to the `models/santa-wrap.mzn` file) and parameters for
driving MiniZinc (i.e., which solver to use, whether there is a timeout or to
search forever, but also, how to create intermediary `.json` input files that
play the role of the `models/santa-wrap001.dzn` for each different possible
input).  For this project, we use the `simpleMiniZinc` constructor to get a
`Minizinc`that also adds some `Hashable` constraint on the input type so that
we can generate filenames.  Fortunately, `Hashable` is generically-derivable as
well so little boilerplate is incured.

The `runLastMinizincJSON` functions runs for up to the timeout provided in the
MiniZinc option and return the latest decisions (if there is any).  A more
general function `runMinizincJSON` exists. This function takes a state-machine
named `ResultHandler` to handle individual results. Indeed, we can ask MiniZinc
to return multiple solutions, prove that there are no more solutions, and
handling this variety of cases asks for more complicated types.

The underlying type are:

```
data SearchState a
  = Exhausted a
  | Incomplete a
  | Unsatisfiable
  | InternalError String

data ResultHandler obj b
  = ResultHandler
  { handleNext :: b -> SearchState obj -> IO (b, Maybe (ResultHandler obj b))
  }
```

The result handler is parametrized by `obj` which are the decisions that the
solver makes, and the handler is also parametrized by `b` -- an accumulator
state like in a fold. A handler is called each time the search changes
SearchState. The handler is free to do what it wants but must return an updated
state and an updated handler (or an absence of handler to tell the library that
we can stop the search).  Later versions of the library will likely generalize
IO in the handler to some parametric class.

CHANGELOG.md
# Revision history for santa-wrap

## 0.1.0.0 -- YYYY-mm-dd

* First version. Released on an unsuspecting world.


Main.hs
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE TypeApplications #-}
{-# LANGUAGE DeriveFunctor #-}
{-# LANGUAGE GeneralizedNewtypeDeriving #-}
{-# LANGUAGE DerivingStrategies #-}
{-# LANGUAGE RecordWildCards #-}
module Main where

import qualified Data.ByteString.Lazy.Char8 as ByteString
import GHC.Generics (Generic)
import Data.Aeson (FromJSON(..), ToJSON(..), eitherDecode')
import Data.Hashable (Hashable)
import Data.Text (Text)
import Data.Foldable (traverse_)
import qualified Data.Text as Text
import qualified Data.List as List
import System.Environment (getArgs)
import Text.Read (readMaybe)

import Codec.Picture
import Process.Minizinc

type Length = Int
type X = Int
type Y = Int
type Name = Text
type Quantity = Int

data Rectangle = Rectangle {
    rectX :: Length
  , rectY :: Length
  , offX  :: X
  , offY  :: Y
  } deriving (Show, Generic)

instance FromJSON Rectangle

data Wrapping = Wrapping {
    name       :: Name
  , colorNum   :: Int
  , quantity   :: Quantity
  , rectangles :: [ Rectangle ]
  } deriving (Show, Generic)

instance FromJSON Wrapping

-- | How far a Wrapping goes in the X-direction.
extentX :: Wrapping -> Length
extentX wrapping = maximum (0 : [ rectX r + offX r | r <- rectangles wrapping ])

-- | How far a Wrapping goes in the Y-direction.
extentY :: Wrapping -> Length
extentY wrapping = maximum (0 : [ rectY r + offY r | r <- rectangles wrapping ])


printWrapping :: Wrapping -> IO ()
printWrapping wrapping =
    writePng path $ generateImage pixelRenderer maxX maxY
  where
    -- File to store the template.
    path = Text.unpack (name wrapping) <> ".png"

    -- Render a pixel, since we'll scan all pixels we need some utility function to test
    -- whether a pixel falls in a rectangle or not.
    pixelRenderer x y =
      case findRectangle x y of
        Nothing -> PixelRGB8 (fromIntegral x `mod` 32) (fromIntegral y `mod` 32) 128
        Just idx -> pixelColor idx

    -- Tests whether a pixel should be displayed green or not
    findRectangle x y = List.findIndex (rectangleContainsPixel x y) (rectangles wrapping)
    rectangleContainsPixel x y rect =
            let x0 = scale $ offX rect
                x1 = scale $ offX rect + rectX rect
                y0 = scale $ offY rect
                y1 = scale $ offY rect + rectY rect
            in  x0 <= x && x < x1 && y0 <= y && y < y1

    -- bounds and scaling
    maxX = scale $ extentX wrapping
    maxY = scale $ extentY wrapping
    scale t = 10 * t

data Positioned a = Positioned {
    posX :: X
  , posY :: Y
  , item :: a
  } deriving (Show, Functor)

data Layout = Layout {
    positionedWrappings :: [ Positioned Wrapping ]
  } deriving (Show)


pixelColor :: Int -> PixelRGB8
pixelColor idx
  | idx `mod` 10 == 0 = PixelRGB8 0 200 0
  | idx `mod` 10 == 1 = PixelRGB8 200 0 0
  | idx `mod` 10 == 2 = PixelRGB8 0 0 200
  | idx `mod` 10 == 3 = PixelRGB8 200 200 0
  | idx `mod` 10 == 4 = PixelRGB8 0 200 200
  | idx `mod` 10 == 5 = PixelRGB8 200 0 200
  | idx `mod` 10 == 6 = PixelRGB8 100 100 0
  | idx `mod` 10 == 7 = PixelRGB8 0 100 100
  | idx `mod` 10 == 8 = PixelRGB8 100 0 100
  | idx `mod` 10 == 9 = PixelRGB8 100 100 100


printLayout :: FilePath -> Layout -> IO ()
printLayout path layout =
    writePng path $ generateImage pixelRenderer maxX maxY
  where
    pixelRenderer :: Int -> Int -> PixelRGB8
    pixelRenderer x y =
      case findWrapping x y of
        Nothing -> PixelRGB8 (fromIntegral x `mod` 32) (fromIntegral y `mod` 32) 128
        Just w -> pixelColor $ colorNum $ item w

    findWrapping :: Int -> Int -> Maybe (Positioned Wrapping)
    findWrapping x y = List.find (wrappingContainsPixel x y) (positionedWrappings layout)

    wrappingContainsPixel :: Int -> Int -> Positioned Wrapping -> Bool
    wrappingContainsPixel x y wrapping =
        List.any
            (rectangleContainsPixel (posX wrapping) (posY wrapping) x y)
            (rectangles $ item wrapping)

    rectangleContainsPixel :: X -> Y -> Int -> Int -> Rectangle -> Bool
    rectangleContainsPixel posx posy x y rect =
            let x0 = scale $ posx + offX rect
                x1 = scale $ posx + offX rect + rectX rect
                y0 = scale $ posy + offY rect
                y1 = scale $ posy + offY rect + rectY rect
            in  (x0 + margin) <= x && x < (x1 - margin) && (y0 + margin) <= y && y < (y1 - margin)

    -- bounds and scaling
    maxX = scale $ maximum (0 : [ extentX (item w) + posX w | w <- positionedWrappings layout ])
    maxY = scale $ maximum (0 : [ extentY (item w) + posY w | w <- positionedWrappings layout ])
    scale t = 10 * t
    margin = 5

linearLayout :: [ Wrapping ] -> Layout
linearLayout wrappings = Layout $ List.zipWith (\w (x,y) -> Positioned x y w) wrappings xys
  where
    xys :: [(X,Y)]
    xys = List.scanl placeToTheLeft (0,0) wrappings

    placeToTheLeft :: (X, Y) -> Wrapping -> (X, Y)
    placeToTheLeft (x,y) w = (x + extentX w, 0)

newtype ShapeIdx = ShapeIdx Int
  deriving stock Show
  deriving stock Eq
  deriving newtype Hashable
  deriving newtype ToJSON
newtype BlockIdx = BlockIdx Int
  deriving stock Show
  deriving newtype Hashable
  deriving newtype ToJSON

data MinizincSet a = MinizincSet { set :: [ a ] }
  deriving (Show, Generic)
instance Hashable a => Hashable (MinizincSet a)
instance ToJSON a => ToJSON (MinizincSet a)

data Input = Input {
    nBlocks :: Int
  , rect_sizes :: [[Int]]
  , rect_offs :: [[Int]]
  , nShapes :: Int
  , shapes :: [ MinizincSet BlockIdx ]
  , nObjs :: Int
  , object_shapes :: [ ShapeIdx ]
  , x_max :: Int
  , y_max :: Int
  } deriving (Show, Generic)
instance Hashable Input
instance ToJSON Input

data Output = Output {
    coordinates :: [ (X, Y) ]
  }
  deriving (Show, Generic)
instance FromJSON Output

convertInput :: Length -> Length -> [ Wrapping ] -> (Input, Output -> Layout)
convertInput sizeX sizeY wrappings = (Input {..}, layoutOutput)
  where
    x_max = sizeX
    y_max = sizeY

    nBlocks = length indexedRects
    nShapes = length wrappings

    -- flatten all rectangles from all wrapping, add some indices
    indexedWrappings :: [(ShapeIdx, Wrapping)]
    indexedWrappings = zip (fmap ShapeIdx [1..]) wrappings

    flattenedWrappings :: [(ShapeIdx, Wrapping)]
    flattenedWrappings = mconcat [ replicate (quantity w) (shapeIdx, w) | (shapeIdx, w) <- indexedWrappings ]

    object_shapes = fmap fst flattenedWrappings
    nObjs = length flattenedWrappings

    indexedRects :: [(ShapeIdx, BlockIdx, Rectangle)]
    indexedRects = zipWith (\blockIdx (shapeIdx, r) ->  (shapeIdx, blockIdx, r))
      (fmap BlockIdx [ 1.. ])
      [ (shapeIdx, r) | (shapeIdx,w) <- indexedWrappings, r <- rectangles w ]

    rect_sizes = [ [rectX r, rectY r]  | (_,_,r) <- indexedRects ]
    rect_offs = [ [offX r, offY r] | (_,_,r) <- indexedRects ]
    shapes = [ MinizincSet [ blockIdx | (shapeIdx2,blockIdx,_) <- indexedRects, shapeIdx1 == shapeIdx2 ]
             | (shapeIdx1,_) <- indexedWrappings
             ]

    layoutOutput :: Output -> Layout
    layoutOutput (Output coords) = Layout
        $ zipWith (\w (x,y) -> Positioned x y w) [w | (_,w) <- flattenedWrappings] coords

main :: IO ()
main = do
  args <- getArgs
  case args of
     ["parts"] -> mainParts
     ["linear-layout", path] -> mainLinearLayout path
     ["minizinc-layout", path, xsize, ysize ] -> do
        case (,) <$> readMaybe xsize <*> readMaybe ysize of
         Just (x,y) -> mainMinizincLayout path x y
         Nothing -> mainUsage
     _ -> mainUsage
  where
    mainUsage = do
      putStrLn "parts | linear-layout <path> | minizinc-layout <path> <width> <height>"
    mainParts = do
      wrappings <- eitherDecode' @[Wrapping] <$> ByteString.getContents
      case wrappings of
        Left err -> print err
        Right xs -> traverse_ printWrapping xs
      putStrLn "done!"
    mainLinearLayout path = do
      wrappings <- eitherDecode' @[Wrapping] <$> ByteString.getContents
      case wrappings of
        Left err -> print err
        Right xs -> do
          let items = mconcat [ replicate (quantity w) w | w <- xs ]
          let layout = linearLayout $ items
          printLayout path layout
      putStrLn "done!"
    mainMinizincLayout path xsize ysize = do
      wrappings <- eitherDecode' @[Wrapping] <$> ByteString.getContents
      case wrappings of
        Left err -> print err
        Right xs -> do
           let (input, convertOutput) = convertInput xsize ysize xs
           let mzn = simpleMiniZinc @Input @Output "models/santa-wrap.mzn" 100000 Chuffed
           out <- runLastMinizincJSON mzn input
           case fmap convertOutput out of
             Nothing -> putStrLn "no layout found!"
             Just layout -> do
               printLayout path layout
               putStrLn "done!"
